src/Catalyst.Core.Modules.Web3/Web3EthApiExtensions.cs

```diff
@@ -0,0 +1,194 @@
+            };
+        }
+        
+        public static TransactionForRpc ToTransactionForRpc(this IWeb3EthApi api, DeltaWithCid deltaWithCid, int transactionIndex)
+        {
+            var (delta, deltaCid) = deltaWithCid;
+            var publicEntry = delta.PublicEntries[transactionIndex];
+            var deltaNumber = delta.DeltaNumber;
+
+            return new TransactionForRpc
+            {
+                GasPrice = publicEntry.GasPrice.ToUInt256(),
+                BlockHash = deltaCid,
+                BlockNumber = (UInt256) deltaNumber,
+                Nonce = publicEntry.Nonce,
+                To = ToAddress(publicEntry.ReceiverAddress),
+                From = ToAddress(publicEntry.SenderAddress),
+                Value = publicEntry.Amount.ToUInt256(),
+                Hash = publicEntry.GetHash(api.HashProvider),
+                Data = publicEntry.Data.ToByteArray(),
+                R = new byte[0],
+                S = new byte[0],
+                V = UInt256.Zero,
+                Gas = publicEntry.GasLimit,
+                TransactionIndex = (UInt256) transactionIndex
+            };
+        }
+
+        static Address ToAddress(ByteString address)
+        {
+            if (address == null || address.IsEmpty)
+            {
+                return null;
+            }
+
+            return new Address(address.ToByteArray());
+        }
+
+        public static CallOutputTracer CallAndRestore(this IWeb3EthApi api, TransactionForRpc transactionCall, DeltaWithCid deltaWithCid)
+        {
+            var parentDelta = deltaWithCid.Delta;
+            Keccak root = parentDelta.StateRoot.ToKeccak();
+
+            if (transactionCall.Gas == null)
+            {
+                transactionCall.Gas = parentDelta.GasLimit;
+            }
+
+            var publicEntry = api.ToPublicEntry(transactionCall, root);
+
+            var newDelta = deltaWithCid.CreateOneOffDelta(publicEntry);
+
+            CallOutputTracer callOutputTracer = new CallOutputTracer();
+
+            api.StateProvider.StateRoot = root;
+            api.Executor.CallAndReset(newDelta, callOutputTracer);
+            api.StateProvider.Reset();
+            api.StorageProvider.Reset();
+            return callOutputTracer;
+        }
+    }
+
+    public struct DeltaWithCid
+    {
+        public Delta Delta;
+        public Cid Cid;
+
+        /// <summary>
+        /// Creates a delta for one off execution.
+        /// </summary>
+        public Delta CreateOneOffDelta(PublicEntry publicEntry)
+        {
+            Delta newDelta = Delta.Clone();
+
+            newDelta.PreviousDeltaDfsHash = Cid.ToArray().ToByteString();
+            newDelta.CoinbaseEntries.Clear();
+            newDelta.ConfidentialEntries.Clear();
+            newDelta.PublicEntries.Clear();
+            newDelta.PublicEntries.Add(publicEntry);
+            return newDelta;
+        }
+
+        public void Deconstruct(out Delta delta, out Cid cid)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Consider refactoring this method in order to remove the need for this 'out' modifier.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4979613)